### ISSUE TITLE
Elision lint should not do anything for zero input lifetimes

### ISSUE THREAD

**Issue by Manishearth** [Manishearth/rust-clippy#291]
`pub unsafe fn as_ref<'a>(&self) -> Option<&'a T> where T: Sized` 

This is not elidable, this is explicitly _avoiding_ elision.

If we have zero input lifetimes, but there are nonzero positions for input lifetimes, we should bail. Same goes for having zero output lifetimes but nonzero positions for output lifetimes.

cc @birkenfeld

**Comment by Manishearth** [comment]
Note that it might be nontrivial to count "positions for input lifetimes" via the AST since things like `Iter<T>` (actually `Iter<'elided, T>` aren't visible in the AST). 

It might work to get the `Ty`, [`walk`](http://manishearth.github.io/rust-internals-docs/rustc/middle/ty/struct.TyS.html#method.walk) it, and look for:

 - `TyRef`s
 - `TyStruct`s or `TyEnum`s with nonempty `Substs`

So, the check becomes:

If it is flagged for input-output lifetime elision, ensure that the input space does not have any positions for lifetimes.

I think that's the only case where you can both avoid elision (to get a different result) and still have it compile. Not sure.

**Comment by Manishearth** [comment]
We also have issues whenever there is a trait object involved:

```rust
fn foo<'a>(&'a self) -> &'a Trait {

}
```

This is not elidable

**Comment by birkenfeld** [comment]
I think the first part is just the match here: https://github.com/Manishearth/rust-clippy/blob/master/src/lifetimes.rs#L129

This faulty arm can just be removed. (Line 128 should still be good though, yes?)
